### PR TITLE
Add printer gadget and example

### PR DESCRIPTION
**Pull request by tuna-f1sh**
Last one from me I think!

Gadget is fairly simple. Included example based on [prn_example](https://docs.kernel.org/6.12/usb/gadget_printer.html#example-code) in kernel docs. I put some const I used in the example that I think are useful in the lib too; ioctl ID bytes.

### CODE REVIEW

File: examples/printer.rs [hunk 1]
```diff
@@ -0,0 +1,154 @@
+//! Printer example userspace application based on [prn_example](https://docs.kernel.org/6.6/usb/gadget_printer.html#example-code)
+//!
+//! Creates and binds a printer gadget function, then reads data from the sysfs file created by the gadget to stdout. Will exit after printing a set number of pages.
```
Comment (surban) [inline examples/printer.rs:3]:
```suggestion
//! Creates and binds a printer gadget function, then reads data from the device file created by the gadget to stdout. Will exit after printing a set number of pages.
```

File: examples/printer.rs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+use std::io::{self, Read, Write};
+use std::os::unix::io::AsRawFd;
+use std::path::PathBuf;
+
+use usb_gadget::function::printer::{Printer, StatusFlags, GADGET_GET_PRINTER_STATUS, GADGET_SET_PRINTER_STATUS};
+use usb_gadget::{default_udc, Class, Config, Gadget, Id, RegGadget, Strings, GADGET_IOC_MAGIC};
+
+// Printer read buffer size, best equal to EP wMaxPacketSize
+const BUF_SIZE: usize = 512;
+// Printer sysfs path - 0 assumes we are the only printer gadget!
```
Comment (surban) [inline examples/printer.rs:15]:
```suggestion
// Printer device path - 0 assumes we are the only printer gadget!
```

File: examples/printer.rs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+const SYSFS_PATH: &str = "/dev/g_printer0";
```
Comment (surban) [inline examples/printer.rs:16]:
```suggestion
const DEV_PATH: &str = "/dev/g_printer0";
```

File: examples/printer.rs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    let g_printer = create_printer_gadget().map_err(|e| {
+        eprintln!("Failed to create printer gadget: {:?}", e);
+        e
+    })?;
+    println!("Printer gadget created: {:?}", g_printer.path());
+
+    // wait for sysfs device to create
+    let mut count = 0;
+    let mut sysfs_path = None;
+    println!("Attempt open sysfs path: {}", SYSFS_PATH);
```
Comment (surban) [inline examples/printer.rs:124]:
```suggestion
    println!("Attempt open device path: {}", DEV_PATH);
```

File: src/function/printer.rs [hunk 5]
```diff
@@ -0,0 +1,97 @@
+//! Printer function.
+//!
+//! The Linux kernel configuration option `CONFIG_USB_CONFIGFS_F_PRINTER` must be enabled.
+//!
+//! A sysfs file at /dev/g_printerN will be created for each instance of the function, where N instance number. See './examples/printer.rs' for an example.
```
Comment (surban) [inline src/function/printer.rs:5]:
```suggestion
//! A device file at /dev/g_printerN will be created for each instance of the function, where N instance number. See 'examples/printer.rs' for an example.
```

File: src/function/printer.rs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+/// Get printer status ioctrl ID
+pub const GADGET_GET_PRINTER_STATUS: u8 = 0x21;
+/// Set printer status ioctrl ID
+pub const GADGET_SET_PRINTER_STATUS: u8 = 0x22;
+
+bitflags! {
+    #[derive(Clone, Copy, Debug)]
+    #[non_exhaustive]
+    /// Printer status flags https://www.usb.org/sites/default/files/usbprint11a021811.pdf
```
Comment (surban) [inline src/function/printer.rs:23]:
```suggestion
    /// [Printer status flags](https://www.usb.org/sites/default/files/usbprint11a021811.pdf)
```

File: src/function/printer.rs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+pub struct Printer {
+    dir: FunctionDir,
+}
+
+impl Printer {
+    /// Creates a new USB human interface device (PRINTER) builder.
+    pub fn builder() -> PrinterBuilder {
+        PrinterBuilder { pnp_string: None, qlen: None }
+    }
+
```
Comment (surban) [inline src/function/printer.rs:92]: A `new` function could be useful, since I guess the `pnp_string` is optional.

**Review by surban — CHANGES_REQUESTED** [review]
Minor changes.

**Comment by tuna-f1sh** [comment]
Thanks. I've fed these changes back in.

**Review by surban — APPROVED** [review]

**Comment by surban** [comment]
Thanks. 

Are there any gadgets left unimplemented? ;-)

**Comment by tuna-f1sh** [comment]
Haha, I don't think so! Thanks for your help and support on these though. Learnt some things.

**Pull request merged**